components/datetime/src/provider/neo.rs

```diff
@@ -544,6 +545,131 @@ pub struct DateTimePatternV1<'data> {
+}
+
+/// Bit-packed [`ULE`] variant of [`SkeletonDataIndex`].
+#[derive(Debug, Copy, Clone, ULE)]
+#[repr(transparent)]
+pub struct SkeletonDataIndexULE([u8; 2]);
+
+impl AsULE for SkeletonDataIndex {
+    type ULE = SkeletonDataIndexULE;
+
+    fn to_unaligned(self) -> Self::ULE {
+        let mut flags = 0;
+        flags |= (self.has_long as u8) << 7;
+        flags |= (self.has_medium as u8) << 6;
+        flags |= (self.has_plurals as u8) << 5;
+        SkeletonDataIndexULE([flags, self.index])
+    }
+
+    fn from_unaligned(unaligned: Self::ULE) -> Self {
+        let [flags, index] = unaligned.0;
+        // TODO: `flags` could have more bits set, but we don't check that here.
+        SkeletonDataIndex {
+            has_long: (flags & (1 << 7)) != 0,
+            has_medium: (flags & (1 << 6)) != 0,
+            has_plurals: (flags & (1 << 5)) != 0,
+            index,
+        }
+    }
+}
+
+#[icu_provider::data_struct(
+    marker(
+        GregorianDateNeoSkeletonPatternsV1Marker,
+        "datetime/patterns/gregory/date_skeleton@1"
+    ),
+    marker(TimeNeoSkeletonPatternsV1Marker, "datetime/patterns/time_skeleton@1")
+)]
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_datetime::provider::neo),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[allow(missing_docs)] // TODO
+pub struct PackedSkeletonDataV1<'data> {
+    // len = 12 for time, 17 for date
+    #[allow(missing_docs)] // TODO
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub indices: ZeroVec<'data, SkeletonDataIndex>,
+    // TODO: This should support plurals
+    #[allow(missing_docs)] // TODO
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub patterns: VarZeroVec<'data, PatternULE>,
+}
+
+impl<'data> PackedSkeletonDataV1<'data> {
+    pub(crate) fn get_for_date_skeleton(
+        &self,
+        sk: NeoDateSkeleton,
+    ) -> Option<&runtime::PatternULE> {
+        self.get_for_discriminant(sk.components.discriminant())
+    }
+    pub(crate) fn get_for_time_skeleton(
+        &self,
+        sk: NeoTimeSkeleton,
+    ) -> Option<&runtime::PatternULE> {
+        self.get_for_discriminant(sk.components.discriminant())
+    }
+    fn get_for_discriminant(&self, discriminant: u8) -> Option<&runtime::PatternULE> {
+        self.indices
+            .get(discriminant as usize)
+            .and_then(|index_info| self.patterns.get(index_info.index as usize))
+    }
+}
+
+#[icu_provider::data_struct(marker(
+    DateTimeSkeletonPatternsV1Marker,
+    "datetime/patterns/datetime_skeleton@1"
+))]
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_datetime::provider::neo),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[yoke(prove_covariance_manually)]
+#[allow(missing_docs)] // TODO
+pub struct DateTimeSkeletonsV1<'data> {
```

Comment:
nit: should we call this something else since it's about exceptions / "weird" skeletons?